rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -29,26 +29,79 @@ and we will not allow them to be provisioned. Although other resources might be
-with Teleport's fine-grained RBAC definitions for Kubernetes. This will make enrolling and starting using Kubernetes clusters
-into Teleport a much simpler process, since user will be able to setup RBAC completely in Teleport if they want. We will also be able
+### Improving Day 1 experience - Adding capability to manage Kubernetes RBAC from Teleport. 
+
+Currently users rely on `kubernetes_groups` and/or `kubernetes_users` fields of the role definition when setting up
+access for Kubernetes. This requires roles/bindings to be created by the client in Kubernetes cluster independently from the 
+RBAC setup in the Teleport. These roles/bindings also should be kept in sync with the Teleport RBAC side,
+which creates additional hurdles for clients, especially the new ones, when using Kubernetes access. 
+To improve this situation we will introduce auto-provisioning of the Kubernetes RBAC based on 
+the Teleport roles, which means users will be able to define required permissions completely in the Teleport, and under the hood
+we will keep them in sync transparently for the users.
+
+We will add a new field to the role definition - `kubernetes_permissions`. That field will contain desired Kubernetes RBAC permissions
+for that role and will be similar in structure to the native Kubernetes RBAC:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_permissions:
+      namespaces: ["namespace1", "namespace2"]
+      rules:
+      - resources: ["pod", "pod/exec"]
+        verbs: ["get", "list", "create"]
+      - resources: ["deployments"]
+        apiGroups: ["app"]
+        verbs: ["*"]
+      - resources: ["secrets"]
+        apiGroups: [""]
+        resourceNames: ["secret1", "secret2"]
+        verbs: ["get", "list""]
+```
+
+Presence of this field will signal that user wants Teleport to auto-provision these permissions into the Kubernetes clusters.
+Teleport will translate permissions defined in that field into Kubernetes RBAC definitions and will automatically create roles and bindings
+on the target Kubernetes clusters (as specified by the labels). If `namespaces` field includes '*' Teleport will create ClusterRole/ClusterRoleBinding
+pair on the Kubernetes cluster, otherwise Role/Binding pair will be created in the each listed namespace. 
+
+Subject for the bindings will be a group named `teleport:*RoleName*`, e.g. for a Teleport role `kube-role` the resulting 
+group name will be `teleport:kube-role`. This will allow Kube service to setup correct impersonation group headers based on the roles available to the user.
+
+If `kubernetes_permissions` field is set in the `allow` part of the Teleport role, fields `kubernetes_resources`, `kubernetes_users` and `kubernetes_groups` 
+should not be set, we will return an error if user will try to create a role mixing those fields. In the future we might relax this rule.
```

Comment:
Did you mean `access` role (`editor` is about editing Teleport resources)?
I think we probably shouldn't put `kubernetes_permissions` in there, `kubernetes_resources` are present there yes, but this is more because as you said we populate these if you have `kubernetes_labels`/expressions set on a role and `access` role has wildcard labels for Kubernetes so you could see all clusters. If we add `kubernetes_permissions` by default to the `access` role it means we will always create `teleport:access` cluster role, which might be unexpected for users who didn't plan to use kube auto-provisioning, although it might an interesting possibility to explore 🤔 - currently `access` role doesn't actually give you access to the Kubernetes resources, because you need to have some groups as well, if we add `kubernetes_permission` it will mean you would have actual access to everything inside the Kubernetes clusters. Maybe start smaller and add this later?

Generally idea about forbidding mixing of the fields is to lower confusion among the users - basically you have two modes: 
1. Use groups/users and `kubernetes_resources` - user is responsible for creating Kubernetes RBAC in that case.
2. Use `kubernetes_permissions` - Teleport is responsible for Kubernetes RBAC in that case.

Maybe we can get away with the default wildcard rule, but I think it would be better if users didn't even see `kubernetes_resources` field if they are using `kubernetes_permissions`

Reply:
> Maybe we can get away with the default wildcard rule, but I think it would be better if users didn't even see kubernetes_resources field if they are using kubernetes_permissions

I agree but I think this is not how it's implemented today. Hwne `kubernetes_resources` is not set, the wildcar item is added to the list and this is what you get when you `tctl get role/...` .

Reply:
I think actually we should implicitly convert `kubernetes_permissions` to `kubernetes_resources` when doing permissions check on the Teleport side. Doing wildcard kubernetes resources can mess things up if user has other roles, that rely on `kubernetes_resources`. I'm not even sure us defaulting to wildcard resources for a role is the right behaviour - imagine situation, user has `kuberole1` that has powerful wide-access `kubernetes_groups` but then is limited by the `kubernetes_resources`. Now admin wants to give access to other Kubernetes clusters, so they create role with additional `kubernetes_labels` so the user can see the clusters, but they forget to add `kubernetes_resources`; we then create wildcard resources for them and now the user with that role has access to everything that powerful `kubernetes_groups` allows.

So we either need to forbid mixing of `kubernetes_permissions` roles and `kubernetes_resources`, which I think is not a good UX; or we translate permissions to `kubernetes_resources` analog just-in-time when doing permissions checks and we don't create default wildcard `kubernetes_resources` if `kubernetes_permissions` are present.

Reply:
> And we don't create default wildcard kubernetes_resources if kubernetes_permissions are present.

I missed this part initially when reading the comment. I think doing conditional defaults like this would cause inconsistencies and surprise behaviour changes.

Case 1:
- I create a simple role v7 without kube_resources nor kube_permissions
- the resource in the backend has the wildcard
- I tctl edit the resource to add kube_permission
- I now have a resource with both kube_permission and wildcard kube_resources

Case 2:
- I create a simple role v7 without kube_resources nor kube_permissions
- the resource in the backend has the wildcard
- I add the kube_permission field to the local manifest and `tctl create -f`
- I now have a resource with both kube_permission and without kube_resources

Such "smart" default behaviours are a big issue in Teleport today because we evaluate defaults before writing the resource to the backend, so they are persisted. We end up with an hysteresis and this messes with people, and especially infrastructure as code tooling.

Reply:
In the way I propose you won't be able to tctl edit resource to add `kube_permissions` without explicitly deleting `kube_resources` wildcard. In case 2 as I understand you mean that you would overwrite resource, which to me also seems like a deliberate action - if you first create role without anything and then you also try to create it with a new yaml file it's a bit weird, but you end up with what you wrote to the yaml file. If you first get yaml of the resource and edit it to add `kube_permissions` you again need to explicitly delete `kube_permissions` from there or it would return an error.

And as I mentioned, I'm not sure that creating wildcard by default for `kube_resources` is the right thing to do - a role with wildcard `kube_resources` is a dangerous role, it can screw up permissions if combined with another role that has wide-access `kubernetes_groups`. Though removing it would be a breaking change. But it's not linked to `kube_permissions` directly if we don't allow mixing of the fields, and it's something we can think about later.